### PR TITLE
Added cgroups v2 CPU implementation.

### LINKED ISSUE #104: cgroup2: check list for v2 controller implementations
## Urgent (required to get CRI and Moby work)
- [x] pids: #103
- [x] cpu: #107
- [x] memory (`memory.stats`): #108 @Zyqsempai
- [x] `memory.current`, `memory.swap.current`, and other memory stat files  https://www.kernel.org/doc/Documentation/cgroup-v2.txt : #117 
- [x] OOM event FD: #115
- [x] io: #112
- [x] devices (BPF): #116
- [x] `ToggleControllers()` (issue: https://github.com/containerd/cgroups/issues/121): #124

Metrics functions are urgent.
`Update()` functions are not necessary for CRI and Moby.

## Non-urgent
- [x] `NewSystemd()` and `LoadSystemd()`: #125 
- [x] freezer: #103
- [ ] perf_event
- [x] cpuset: #111
- [x] rdma: #119 
- [x] [hugetlb](https://www.spinics.net/lists/cgroups/msg24003.html) (to appear in kernel 5.XX)

### PR DESCRIPTION
**Pull request by Zyqsempai**
Partially fixes #104 
Added cgroups v2 CPU implementation.

Signed-off-by: bpopovschi <zyqsempai@mail.ru>

### CODE REVIEW

**Comment by Zyqsempai** [comment]
@AkihiroSuda Can you add some acceptance criteria or checkpoints what should be done, for easier testing and validation for this and next PR's.

File: v2/cpu.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+		}
+	}
+	return nil
+}
+
+func (c *cpuController) Update(g GroupPath, resources *specs.LinuxResources) error {
+	return c.Create(g, resources)
+}
+
+func (c *cpuController) Stat(g GroupPath, stats *v1.Metrics) error {
```
Comment (AkihiroSuda) [inline v2/cpu.go]: statsv2.Metrics

**Comment by AkihiroSuda** [comment]
Please add to https://github.com/containerd/cgroups/blob/6fb48e16d8117807adf676f7c735f4f553d19fde/v2/utils.go#L32

**Comment by AkihiroSuda** [comment]
wrt tests I think we can add metrics code to cmd/cgroups-playground and call it a day.

In future we should have proper Go tests

**Comment by Zyqsempai** [comment]
@AkihiroSuda The problem is that https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L297 this config still not updated to v2.
Should I open an issue and corresponding PR to `runtime-spec`?
For example how it's done in `runc` https://github.com/opencontainers/runc/blob/master/libcontainer/configs/cgroup_linux.go#L123

**Comment by AkihiroSuda** [comment]
No, you can just convert
https://github.com/containers/crun/blob/master/crun.1.md#cgroup-v2

**Comment by Zyqsempai** [comment]
@AkihiroSuda PTAL

File: cmd/cgroups-playground/main.go [hunk 2]
```diff
@@ -54,5 +57,30 @@ func xmain() error {
 	for i, s := range subsystems {
 		logrus.Infof("Subsystem %d: %q", i, s.Name())
 	}
+
+	cpuCgroup, err := v2.NewCpu(unifiedMountpoint)
+	if err != nil {
+		return err
+	}
+	cpuGroupPath := v2.GroupPath(filepath.Join(unifiedMountpoint, "cpu"))
```
Comment (AkihiroSuda) [inline cmd/cgroups-playground/main.go]: Creating "cpu" under "/sys/fs/cgroup" is confusing, it looks as if mixing up v1 and v2

File: v2/cpu.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+func (c *cpuController) path(g GroupPath) string {
+	return filepath.Join(c.unifiedMountpoint, string(g))
+}
+
+func (c *cpuController) Create(g GroupPath, resources *specs.LinuxResources) error {
+	if err := os.MkdirAll(c.path(g), defaultDirPerm); err != nil {
+		return err
+	}
+	if cpuShares := resources.CPU.Shares; cpuShares != nil {
+		convertedWeight := (1 + ((*cpuShares-2)*9999)/262142)
```
Comment (AkihiroSuda) [inline v2/cpu.go]: please add link to crun man page for these magic numbers

File: v2/stats/metrics.proto [hunk 4]
```diff
@@ -2,13 +2,34 @@ syntax = "proto3";
...
+message CPUUsage {
+	// values in nanoseconds
+	uint64 total = 1;
+	uint64 kernel = 2;
+	uint64 user = 3;
+	repeated uint64 per_cpu = 4 [(gogoproto.customname) = "PerCPU"];
+
+}
+
+message Throttle {
```
Comment (AkihiroSuda) [inline v2/stats/metrics.proto]: probably better to comment out if this is not going to be implemented in this PR

**Comment by AkihiroSuda** [comment]
thanks, just a couple of nits

File: cmd/cgroups-playground/main.go [hunk 5]
```diff
@@ -54,5 +57,30 @@ func xmain() error {
...
+	var period, shares uint64 = 1000, 5000
+	resources := specs.LinuxResources{
+		CPU: &specs.LinuxCPU{Period: &period, Shares: &shares},
+	}
+	err = cpuCgroup.Create(cpuGroupPath, &resources)
```
Comment (AkihiroSuda) [inline cmd/cgroups-playground/main.go]: defer remove

File: cmd/cgroups-playground/main.go [hunk 6]
```diff
@@ -54,5 +57,37 @@ func xmain() error {
...
+	stats := stats2.Metrics{
+		CPU: &stats2.CPUStat{
+			Usage: &stats2.CPUUsage{},
+		},
+	}
+	err = cpuCgroup.Stat(cpuGroupPath, &stats)
+	if err != nil {
+		return err
+	}
+	logrus.Infof("CPU usage stats: usage in kernel mode - %d", stats.CPU.Usage.Kernel)
```
Comment (AkihiroSuda) [inline cmd/cgroups-playground/main.go]: This isn't interesting without adding process to the current group.
Why not just use the current group?

Comment (Zyqsempai) [reply cmd/cgroups-playground/main.go]: What do you mean, by using current group?

Comment (AkihiroSuda) [reply cmd/cgroups-playground/main.go]: `g, err := v2.PidGroupPath(pid)` in the top of the function

Comment (Zyqsempai) [reply cmd/cgroups-playground/main.go]: make sense, ok.

File: cmd/cgroups-playground/main.go [hunk 7]
```diff
@@ -54,5 +55,36 @@ func xmain() error {
...
+	}
+	var period, shares uint64 = 1000, 5000
+	resources := specs.LinuxResources{
+		CPU: &specs.LinuxCPU{Period: &period, Shares: &shares},
+	}
+	err = cpuCgroup.Create(g, &resources)
+	if err != nil {
+		return err
+	}
+	defer removePath(string(g))
```
Comment (AkihiroSuda) [inline cmd/cgroups-playground/main.go]: this one is no longer needed

**Comment by crosbymichael** [comment]
One thing before we go farther on this. Should v2 be simpler where we don't have separate files/controllers, cpu.go/memory.go but a single manager.go with all the settings?

**Comment by AkihiroSuda** [comment]
Good idea, could you open PR? (maybe after merging this one)

**Comment by crosbymichael** [comment]
Ya, I think we should merge then one and then i can work on a unified manager :)

**Comment by crosbymichael** [comment]
LGTM

@AkihiroSuda after this last edit, does it still LGTM for you?

**Pull request merged**